elkscmd/man/man1/telnet.1
.TH TELNET 1
.SH NAME
telnet \- user interface to the TELNET protocol
.SH SYNOPSIS
.B telnet
[
.I -e
<esc>
-I 
]
.I host
[
.I port
]
.SH DESCRIPTION
.B telnet
is a small and simple implementation of the client side of the TELNET protocol,
used to communicate interactively with another Internet or LAN host.
.PP
Once a connection has been opened, all
text typed is immediately sent to the remote host for processing.
.PP
A hung session may be terminated using the
.I escape
character, '^]' by default, which may be set via the 
.I -e
option. The new escape character may be specified as a single character optionally
preceded by a caret ('^'), which will be converted to the corresponding
 control character, as in '^A' becoming 'control-A' etc.  
.PP
.SH BUGS
.PP
There is no adequate way for dealing with flow control.


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/inet/telnet/telnet.c
/*
 * Telnet for ELKS - TLVC
 *
 * Based on minix telnet client.
 * (c) 2001 Harry Kalogirou <harkal@rainbow.cs.unipi.gr>
 *
 */

#include <sys/types.h>
#include <sys/ioctl.h>
#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <termios.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <netinet/in.h>
#include <arpa/inet.h>

//#define DEBUG 1
//#define RAWTELNET             /* set in telnet and telnetd for raw telnet without IAC*/

#define ESCAPE      (']'&0x1f)  /* = ^] escape session, effectively terminate */
#define BUFSIZE		1500

/* telnet protocol */
#define IAC		255
#define IAC_SE		240
#define IAC_NOP		241
#define IAC_DataMark	242
#define IAC_BRK		243
#define IAC_IP		244
#define IAC_AO		245
#define IAC_AYT		246
#define IAC_EC		247
#define IAC_EL		248
#define IAC_GA		249
#define IAC_SB		250
#define IAC_WILL	251
#define IAC_WONT	252
#define IAC_DO		253
#define IAC_DONT	254

#define OPT_ECHO	1
#define OPT_SUPP_GA	3
#define OPT_TERMTYPE	24

#define TERMTYPE_SEND	1
#define TERMTYPE_IS	0

#define FALSE	0
#define TRUE	(!(FALSE))

#ifdef __linux__
int DO_echo= TRUE;
int DO_echo_allowed= TRUE;
int WILL_terminal_type= FALSE;
int WILL_terminal_type_allowed= TRUE;
int DO_suppress_go_ahead= TRUE;
int DO_suppress_go_ahead_allowed= TRUE;
#else
int DO_echo= FALSE;
int DO_echo_allowed= TRUE;
int WILL_terminal_type= FALSE;
int WILL_terminal_type_allowed= TRUE;
int DO_suppress_go_ahead= FALSE;
int DO_suppress_go_ahead_allowed= TRUE;
#endif

#if DEBUG
#define where() (fprintf(stderr, "%s %d:", __FILE__, __LINE__))
#endif

static int tcp_fd;
static char *term_env;
static struct termios def_termios;
static int writeall (int fd, char *buffer, int buf_size);
static int process_opt (char *bp, int count);
static int escape = ESCAPE;

void finish()
{
	int nonblock = 0;
	ioctl(0, FIONBIO, &nonblock);
	tcsetattr(0, TCSANOW, &def_termios);
	exit(0);	
}

static void keybd(void)
{
	int count;
	char buffer[BUFSIZE];

		count = read(0, buffer, sizeof(buffer));
		if (count <= 0 || buffer[0] == escape) {
			char *msg = NULL;
			if (count <= 0) msg = "read error, ";
			fprintf(stderr, "\r\ntelnet: %ssession terminated\r\n", msg);
			finish();
			return;
		}
#if DEBUG
 { where(); fprintf(stderr, "writing %d bytes\r\n", count); }
#endif
		count = write(tcp_fd, buffer, count);
		if (count < 0)
		{
			perror("Connection closed");
			finish();
		}
}

static void scrn(void)
{
	char *bp, *iacptr;
	int count, optsize;
	char buffer[BUFSIZE];

		count = read(tcp_fd, buffer, sizeof(buffer));
#if DEBUG
 { where(); fprintf(stderr, "read %d bytes\r\n", count); }
#endif
		if (count > (int)sizeof(buffer)) { //FIXME buffer overflow from inet_read
			printf("\r\nTELNET BUFFER OVERFLOW\r\n");
			finish();
		}
		if (count <= 0) {
			if (count < 0)
				perror("Read socket");
			printf("\r\nConnection closed\r\n");
			finish();
		}
#ifdef RAWTELNET
		write(1, buffer, count);
#else
		bp= buffer;
		do
		{
			iacptr= memchr (bp, IAC, count);
			if (!iacptr) {
				write(1, bp, count);
				count= 0;
				return;
			}
			if (iacptr && iacptr>bp) {
#if DEBUG
 { where(); fprintf(stderr, " ptr-bp= %d\r\n", iacptr-bp); }
#endif
				write(1, bp, iacptr-bp);
				count -= (iacptr-bp);
				bp= iacptr;
				continue;
			}
			if (iacptr) {
				optsize= process_opt(bp, count);
#if DEBUG
 { where(); fprintf(stderr, "process_opt(...)= %d\r\n", optsize); }
#endif
				if (optsize<0)
					return;
assert (optsize);
				bp += optsize;
				count -= optsize;
			}
		} while (count);
#endif
}

static int usage(void)
{
	printf("Usage: telnet [-e esc_char] host <port>\n");
	return 1;
}

int main(int argc, char **argv)
{
	unsigned short port;
	struct sockaddr_in locadr, remadr;
	int nonblock;
	ipaddr_t ipaddr;

	if (argc < 2)
		return usage();

	if (*argv[1] == '-') {
		if (*(argv[1]+1) != 'e')
			return usage();
		if (*argv[2] != '^')
			escape = *argv[2];
		else
			escape = *(argv[2]+1);
		escape = (escape&0xdf) - '@';
		argv += 2;
	}

	tcgetattr(0, &def_termios);
	signal(SIGINT, finish);

	tcp_fd = socket(AF_INET, SOCK_STREAM, 0);
	if (tcp_fd < 0) {
		perror("telnet");
		return 1;
	}
	
	locadr.sin_family = AF_INET;
	locadr.sin_port = PORT_ANY;
	locadr.sin_addr.s_addr = INADDR_ANY;

#if 0	/* test code to force RST on telnet exit*/
{
	struct linger l;

	l.l_onoff = 1;
	l.l_linger = 0;		/* send RST on close */
	setsockopt(tcp_fd, SOL_SOCKET, SO_LINGER, &l, sizeof(l));
}
#endif
	if (bind(tcp_fd, (struct sockaddr *)&locadr, sizeof(struct sockaddr_in)) < 0) {
		perror("bind");
		return 1;
	}

	ipaddr = in_gethostbyname(argv[1]);
	if (!ipaddr) {
		perror(argv[1]);
		return 1;
	}
	port = argv[2] ? atoi(argv[2]): 23;

	remadr.sin_family = AF_INET;
	remadr.sin_port = htons(port);
	remadr.sin_addr.s_addr = ipaddr;

	printf("Trying %s (%s:%u)...\n", argv[1], in_ntoa(ipaddr), port);
	if (in_connect(tcp_fd, (struct sockaddr *)&remadr, sizeof(struct sockaddr_in), 10) < 0) {
		perror("Connection failed");
		return 1;
	}
	printf("Connected\n");
	printf("Escape character is '^%c'.\n", escape + '@');

#ifdef RAWTELNET
	{
	struct termios termios;
	tcgetattr(0, &termios);
	termios.c_iflag &= ~(ICRNL|IGNCR|INLCR|IXON|IXOFF);
	termios.c_oflag &= ~(OPOST);
	termios.c_lflag &= ~(ECHO|ECHONL|ICANON|ISIG);
	tcsetattr(0, TCSANOW, &termios);
	}
#endif
	nonblock = 1;
	ioctl(0, FIONBIO, &nonblock);

	for (;;){
		fd_set fdset;
		FD_ZERO(&fdset);
		FD_SET(0, &fdset);
		FD_SET(tcp_fd, &fdset);

		if (select(tcp_fd + 1, &fdset, NULL, NULL, NULL) < 0) {
			perror("select");
			break;
		}

		if (FD_ISSET(tcp_fd, &fdset))
			scrn();
		if (FD_ISSET(0, &fdset))
			keybd();
	}
	finish();
}

#ifndef RAWTELNET
#define next_char(var) \
	if (offset < count) { (var) = bp[offset++]; } \
	else { if (read(tcp_fd, (char *)&(var), 1) != 1) printf("TELNET BAD READ2\n"); exit(1); }

static void do_option(int optsrt)
{
	int result;
	unsigned char reply[3];

	switch (optsrt)
	{
	case OPT_TERMTYPE:
		if (WILL_terminal_type)
			return;
		if (!WILL_terminal_type_allowed) {
			reply[0]= IAC;
			reply[1]= IAC_WONT;
			reply[2]= optsrt;
		} else {
			WILL_terminal_type= TRUE;
			term_env= getenv("TERM");
			if (!term_env)
				term_env= "unknown";
			reply[0]= IAC;
			reply[1]= IAC_WILL;
			reply[2]= optsrt;
		}
		break;
	default:
#if DEBUG
 { where(); fprintf(stderr, "got a DO (%d)\r\n", optsrt); }
#endif
#if DEBUG
 { where(); fprintf(stderr, "WONT (%d)\r\n", optsrt); }
#endif
		reply[0]= IAC;
		reply[1]= IAC_WONT;
		reply[2]= optsrt;
		break;
	}
	result= writeall(tcp_fd, (char *)reply, 3);
	if (result<0)
		perror("write");
}

static void will_option(int optsrt)
{
	int result;
	unsigned char reply[3];

	switch (optsrt)
	{
	case OPT_ECHO:
		if (DO_echo)
			break;
		if (!DO_echo_allowed) {
			reply[0]= IAC;
			reply[1]= IAC_DONT;
			reply[2]= optsrt;
		} else {
			struct termios termios;
			tcgetattr(0, &termios);
			termios.c_iflag &= ~(ICRNL|IGNCR|INLCR|IXON|IXOFF);
			/*termios.c_oflag &= ~(OPOST);*/	/* leave OPOST|ONLCR on*/
			termios.c_lflag &= ~(ECHO|ECHONL|ICANON);	/* leave ISIG on for ^P*/
			termios.c_cc[VINTR] = 0;	/* turn ^C off*/
			termios.c_cc[VSUSP] = 0;	/* turn ^Z off*/
			tcsetattr(0, TCSANOW, &termios);
			DO_echo= TRUE;
			reply[0]= IAC;
			reply[1]= IAC_DO;
			reply[2]= optsrt;
		}
		result = writeall(tcp_fd, (char *)reply, 3);
		if (result < 0)
			perror("write");
		break;

	case OPT_SUPP_GA:
		if (DO_suppress_go_ahead)
			break;
		if (!DO_suppress_go_ahead_allowed) {
			reply[0]= IAC;
			reply[1]= IAC_DONT;
			reply[2]= optsrt;
		} else {
			DO_suppress_go_ahead= TRUE;
			reply[0]= IAC;
			reply[1]= IAC_DO;
			reply[2]= optsrt;
		}
		result = writeall(tcp_fd, (char *)reply, 3);
		if (result < 0)
			perror("write");
		break;

	default:
#if DEBUG
 { where(); fprintf(stderr, "got a WILL (%d)\r\n", optsrt); }
#endif
#if DEBUG
 { where(); fprintf(stderr, "DONT (%d)\r\n", optsrt); }
#endif
		reply[0]= IAC;
		reply[1]= IAC_DONT;
		reply[2]= optsrt;
		result= writeall(tcp_fd, (char *)reply, 3);
		if (result<0)
			perror("write");
		break;
	}
}

static int writeall(int fd, char *buffer, int buf_size)
{
	int result;

	while (buf_size)
	{
		result= write (fd, buffer, buf_size);
		if (result <= 0)
			return -1;
assert (result <= buf_size);
		buffer += result;
		buf_size -= result;
	}
	return 0;
}

static void dont_option(int optsrt)
{
	switch (optsrt)
	{
	default:
#if DEBUG
 { where(); fprintf(stderr, "got a DONT (%d)\r\n", optsrt); }
#endif
		break;
	}
}

static void wont_option(int optsrt)
{
	switch (optsrt)
	{
	default:
#if DEBUG
 { where(); fprintf(stderr, "got a WONT (%d)\r\n", optsrt); }
#endif
		break;
	}
}

static int sb_termtype(char *bp, int count)
{
	unsigned char command, iac, optsrt;
	unsigned char buffer[4];
	int offset, result, ret_value;

	offset= 0;
	next_char(command);
	if (command == TERMTYPE_SEND)
	{
		buffer[0]= IAC;
		buffer[1]= IAC_SB;
		buffer[2]= OPT_TERMTYPE;
		buffer[3]= TERMTYPE_IS;
		result= writeall(tcp_fd, (char *)buffer,4);
		if (result<0){
			ret_value = result;
			goto ret;
		}
		count= strlen(term_env);
		if (!count)
		{
			term_env= "unknown";
			count= strlen(term_env);
		}
		result= writeall(tcp_fd, term_env, count);
		if (result<0){
			ret_value = result;
			goto ret;
		}
		buffer[0]= IAC;
		buffer[1]= IAC_SE;
		result= writeall(tcp_fd, (char *)buffer,2);
		if (result<0){
			ret_value = result;
			goto ret;			
		}

	}
	else
	{
#if DEBUG
 where();
#endif
		fprintf(stderr, "got an unknown command (skipping)\r\n");
	}
	for (;;)
	{
		next_char(iac);
		if (iac != IAC)
			continue;
		next_char(optsrt);
		if (optsrt == IAC)
			continue;
		if (optsrt != IAC_SE)
		{
#if DEBUG
 where();
#endif
		/*	fprintf(stderr, "got IAC %d\r\n", optsrt);*/
		}
		break;
	}	
	ret_value = offset;
ret:
	return ret_value;
}


static int process_opt(char *bp, int count)
{
	unsigned char iac, command, optsrt, sb_command;
	int offset, result;
#if DEBUG
 { where(); fprintf(stderr, "process_opt(bp= 0x%x, count= %d)\r\n",
	bp, count); }
#endif

	offset= 0;
assert (count);
	next_char(iac);
assert (iac == IAC);
	next_char(command);
	switch(command)
	{
	case IAC_NOP:
		break;
	case IAC_DataMark:
//fprintf(stderr, "got a DataMark\r\n");
		break;
	case IAC_BRK:
fprintf(stderr, "got a BRK\r\n");
		break;
	case IAC_IP:
fprintf(stderr, "got a IP\r\n");
		break;
	case IAC_AO:
fprintf(stderr, "got a AO\r\n");
		break;
	case IAC_AYT:
fprintf(stderr, "got a AYT\r\n");
		break;
	case IAC_EC:
fprintf(stderr, "got a EC\r\n");
		break;
	case IAC_EL:
fprintf(stderr, "got a EL\r\n");
		break;
	case IAC_GA:
fprintf(stderr, "got a GA\r\n");
		break;
	case IAC_SB:
		next_char(sb_command);
		switch (sb_command)
		{
		case OPT_TERMTYPE:
#if DEBUG
fprintf(stderr, "got SB TERMINAL-TYPE\r\n");
#endif
			result= sb_termtype(bp+offset, count-offset);
			if (result<0)
				return result;
			else
				return result+offset;
		default:
fprintf(stderr, "got an unknown SB (skiping)\r\n");
			for (;;)
			{
				next_char(iac);
				if (iac != IAC)
					continue;
				next_char(optsrt);
				if (optsrt == IAC)
					continue;
if (optsrt != IAC_SE)
	fprintf(stderr, "got IAC %d\r\n", optsrt);
				break;
			}
		}
		break;
	case IAC_WILL:
		next_char(optsrt);
		will_option(optsrt);
		break;
	case IAC_WONT:
		next_char(optsrt);
		wont_option(optsrt);
		break;
	case IAC_DO:
		next_char(optsrt);
		do_option(optsrt);
		break;
	case IAC_DONT:
		next_char(optsrt);
		dont_option(optsrt);
		break;
	case IAC:
/*fprintf(stderr, "got a IAC\r\n");*/
		break;
	default:
fprintf(stderr, "got unknown command (%d)\r\n", command);
	}
	return offset;
}
#endif /* RAWTELNET*/
